src/impl_owned_array.rs

```diff
@@ -59,3 +69,558 @@ where
+            let new_stride = self.axes().fold(1, |acc, ax| {
+                if ax.axis == axis { acc } else {
+                    let this_ax = ax.len as isize * ax.stride;
+                    if this_ax.abs() > acc { this_ax } else { acc }
+                }
+            });
+            let mut strides = self.strides.clone();
+            strides[axis.index()] = new_stride as usize;
+            strides
+        } else {
+            self.strides.clone()
+        };
+
+        unsafe {
+            // grow backing storage and update head ptr
+            let data_to_array_offset = (self.as_ptr() as isize - self.data.as_ptr() as isize)
+                / std::mem::size_of::<A>() as isize;
+            debug_assert!(data_to_array_offset >= 0);
+
+            self.data.reserve(len_to_append);
+            self.ptr = self.data.as_nonnull_mut().offset(data_to_array_offset);
+
+            // clone elements from view to the array now
+            //
+            // To be robust for panics and drop the right elements, we want
+            // to fill the tail in memory order, so that we can drop the right elements on panic.
+            //
+            // We have: Zip::from(tail_view).and(array)
+            // Transform tail_view into standard order by inverting and moving its axes.
+            // Keep the Zip traversal unchanged by applying the same axis transformations to
+            // `array`. This ensures the Zip traverses the underlying memory in order.
+            //
+            // XXX It would be possible to skip this transformation if the element
+            // doesn't have drop. However, in the interest of code coverage, all elements
+            // use this code initially.
+
+            // Invert axes in tail_view by inverting strides
+            let mut tail_strides = strides.clone();
+            if tail_strides.ndim() > 1 {
+                for i in 0..tail_strides.ndim() {
+                    let s = tail_strides[i] as isize;
+                    if s < 0 {
+                        tail_strides.set_axis(Axis(i), -s as usize);
+                        array.invert_axis(Axis(i));
+                    }
+                }
+            }
+
+            // With > 0 strides, the current end of data is the correct base pointer for tail_view
+            let tail_ptr = self.data.as_end_nonnull();
+            let mut tail_view = RawArrayViewMut::new(tail_ptr, array_shape, tail_strides);
+
+            if tail_view.ndim() > 1 {
+                sort_axes_in_default_order_tandem(&mut tail_view, &mut array);
+                debug_assert!(tail_view.is_standard_layout(),
+                              "not std layout dim: {:?}, strides: {:?}",
+                              tail_view.shape(), tail_view.strides());
+            } 
+
+            // Keep track of currently filled lenght of `self.data` and update it
+            // on scope exit (panic or loop finish).
+            struct SetLenOnDrop<'a, A: 'a> {
```

Comment:
What is the purpose of using a guard? (Why not directly update the length of `self.data`?)

Reply:
We should try that, as long as benchmarks are still ok. I think this pattern became a habit (in Vec, arrayvec etc). From working around aliasing bugs that should now be gone, now that we have noalias enabled for `&mut` (recent rustc + llvm 12 fix).

Reply:
Ah, yes, I see [a comment in `std`](https://github.com/rust-lang/rust/blob/1.51.0/library/alloc/src/vec/mod.rs#L1988-L1990) which explains this kind of thing:

```rust
            // Use SetLenOnDrop to work around bug where compiler
            // may not realize the store through `ptr` through self.set_len()
            // don't alias.
```

I agree that it would be worth trying without the separate guard to see if the performance is the same.

Reply:
What I can see, it's still a worthwhile optimization to do this unfortunately. My rustc version is old `rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)` but still, then current stable is likely affected and the optimization can stay.

What I see - using the benches/append.rs benchmarks, is that the inner loop doesn't optimize very well without the drop guard.

Reply:
Okay. That is unfortunate. I hope this will become unnecessary in the future with compiler improvements.

Reply:
It would be good to add a comment saying that the guard is a performance optimization.